lib/Parse/Lexer.cpp

```diff
@@ -826,10 +826,18 @@ void Lexer::lexDollarIdent() {
     }
   }
 
-  // It's always an error to see a standalone $
   if (CurPtr == tokStart + 1) {
-    diagnose(tokStart, diag::expected_dollar_numeric);
-    return formToken(tok::unknown, tokStart);
+    // It's always an error to see a standalone $ in Swift 3.
+    if (LangOpts.isSwiftVersion3()) {
+      // Offer to replace it with '`$`'.
+      diagnose(tokStart, diag::expected_dollar_numeric);
+      diagnose(tokStart, diag::suggest_backtick_dollar_identifier)
+        .fixItReplaceChars(getSourceLoc(tokStart), getSourceLoc(CurPtr), "`$`");
+      return formToken(tok::unknown, tokStart);
+    } else{
+      // Else form an identifier token.
+      return formToken(tok::identifier, tokStart);
```

Comment:
We should *always* treat it as an identifier, even after complaining, because we should recover as-if the user had accepted the Fix-It and compiled again.